Travis failure (internal/log)

Happens on travis (spuriously). Example https://travis-ci.org/ethereum/go-ethereum/jobs/636889600#L1085 :
```
panic: Log in goroutine after TestServerInboundThrottle has completed

goroutine 9726 [running]:
testing.(*common).logDepth(0xc00010e400, 0xc00057e300, 0x60, 0x3)
	/home/travis/.gimme/versions/go1.12.15.linux.amd64/src/testing/testing.go:634 +0x3a6
testing.(*common).log(...)
	/home/travis/.gimme/versions/go1.12.15.linux.amd64/src/testing/testing.go:614
testing.(*common).Logf(0xc00010e400, 0x99a7cf, 0x2, 0xc000067e98, 0x1, 0x1)
	/home/travis/.gimme/versions/go1.12.15.linux.amd64/src/testing/testing.go:649 +0x7f
github.com/ethereum/go-ethereum/internal/testlog.(*logger).flush(0xc0001d2000)
	/home/travis/gopath/src/github.com/ethereum/go-ethereum/internal/testlog/testlog.go:139 +0xfd
github.com/ethereum/go-ethereum/internal/testlog.(*logger).Trace(0xc0001d2000, 0x9aad13, 0x1c, 0xc000555c40, 0x4, 0x4)
	/home/travis/gopath/src/github.com/ethereum/go-ethereum/internal/testlog/testlog.go:80 +0xc7
github.com/ethereum/go-ethereum/p2p.(*Server).SetupConn(0xc00018a280, 0xa5b060, 0xc0002dc4e0, 0x4, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/ethereum/go-ethereum/p2p/server.go:945 +0x2d8
github.com/ethereum/go-ethereum/p2p.(*Server).listenLoop.func1(0xc00018a280, 0xa5b060, 0xc0002dc4e0, 0xc000100de0)
	/home/travis/gopath/src/github.com/ethereum/go-ethereum/p2p/server.go:915 +0x60
created by github.com/ethereum/go-ethereum/p2p.(*Server).listenLoop
	/home/travis/gopath/src/github.com/ethereum/go-ethereum/p2p/server.go:914 +0x4e7
```

--- a/p2p/server.go
+++ b/p2p/server.go
@@ -864,7 +864,6 @@ func (srv *Server) maxDialedConns() int {
 // listenLoop runs in its own goroutine and accepts
 // inbound connections.
 func (srv *Server) listenLoop() {
-	defer srv.loopWG.Done()
 	srv.log.Debug("TCP listener up", "addr", srv.listener.Addr())
 
 	tokens := defaultMaxPendingPeers
@@ -872,6 +871,12 @@ func (srv *Server) listenLoop() {
 		tokens = srv.MaxPendingPeers
 	}
 	slots := make(chan struct{}, tokens)
+	defer func() {
+		// Wait for a slot. This is to wait for any goroutine(s) doing srv.SetupConn
+		// to complete before exiting
+		<-slots
+		srv.loopWG.Done()
+	}()
 	for i := 0; i < tokens; i++ {
 		slots <- struct{}{}
 	}

--- a/p2p/server.go
+++ b/p2p/server.go
@@ -866,20 +866,24 @@ func (srv *Server) maxDialedConns() int {
 func (srv *Server) listenLoop() {
 	srv.log.Debug("TCP listener up", "addr", srv.listener.Addr())
 
+	// The slots channel limits accepts of new connections.
 	tokens := defaultMaxPendingPeers
 	if srv.MaxPendingPeers > 0 {
 		tokens = srv.MaxPendingPeers
 	}
 	slots := make(chan struct{}, tokens)
-	defer func() {
-		// Wait for a slot. This is to wait for any goroutine(s) doing srv.SetupConn
-		// to complete before exiting
-		<-slots
-		srv.loopWG.Done()
-	}()
 	for i := 0; i < tokens; i++ {
 		slots <- struct{}{}
 	}
+
+	// Wait for slots to be returned on exit. This ensures all connection goroutines
+	// are down before listenLoop returns.
+	defer srv.loopWG.Done()
+	defer func() {
+		for i := 0; i < cap(slots); i++ {
+			<-slots
+		}
+	}()
 
 	for {
 		// Wait for a free slot before accepting.
@@ -896,6 +900,7 @@ func (srv *Server) listenLoop() {
 				continue
 			} else if err != nil {
 				srv.log.Debug("Read error", "err", err)
+				slots <- struct{}{}
 				return
 			}
 			break

